server/src/main/java/org/elasticsearch/index/codec/vectors/es910/hnsw/NeighborArray.java

```diff
@@ -0,0 +1,303 @@
+        } else {
+            insertionPoint = -insertionPoint - 1;
+        }
+        return insertionPoint;
+    }
+
+    private int descSortFindRightMostInsertionPoint(float newScore, int bound) {
+        int start = 0;
+        int end = bound - 1;
+        while (start <= end) {
+            int mid = (start + end) / 2;
+            if (scores.get(mid) < newScore) end = mid - 1;
+            else start = mid + 1;
+        }
+        return start;
+    }
+
+    /**
+     * Find first non-diverse neighbour among the list of neighbors starting from the most distant
+     * neighbours
+     */
+    private int findWorstNonDiverse(UpdateableRandomVectorScorer scorer) throws IOException {
+        int[] uncheckedIndexes = sort(scorer);
+        assert uncheckedIndexes != null : "We will always have something unchecked";
+        int uncheckedCursor = uncheckedIndexes.length - 1;
+        for (int i = size - 1; i > 0; i--) {
+            if (uncheckedCursor < 0) {
+                // no unchecked node left
+                break;
+            }
+            scorer.setScoringOrdinal(nodes.get(i));
+            if (isWorstNonDiverse(i, uncheckedIndexes, uncheckedCursor, scorer)) {
+                return i;
+            }
+            if (i == uncheckedIndexes[uncheckedCursor]) {
+                uncheckedCursor--;
+            }
+        }
+        return size - 1;
+    }
+
+    private boolean isWorstNonDiverse(int candidateIndex, int[] uncheckedIndexes, int uncheckedCursor, RandomVectorScorer scorer)
+        throws IOException {
+        float minAcceptedSimilarity = scores.get(candidateIndex);
+        if (candidateIndex == uncheckedIndexes[uncheckedCursor]) {
+            // the candidate itself is unchecked
+            for (int i = candidateIndex - 1; i >= 0; i--) {
+                float neighborSimilarity = scorer.score(nodes.get(i));
+                // candidate node is too similar to node i given its score relative to the base node
+                if (neighborSimilarity >= minAcceptedSimilarity) {
+                    return true;
+                }
+            }
+        } else {
+            // else we just need to make sure candidate does not violate diversity with the (newly
+            // inserted) unchecked nodes
+            assert candidateIndex > uncheckedIndexes[uncheckedCursor];
+            for (int i = uncheckedCursor; i >= 0; i--) {
+                float neighborSimilarity = scorer.score(nodes.get(uncheckedIndexes[i]));
+                // candidate node is too similar to node i given its score relative to the base node
+                if (neighborSimilarity >= minAcceptedSimilarity) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public int maxSize() {
+        return maxSize;
+    }
+
+    @Override
+    public long ramBytesUsed() {
+        return BASE_RAM_BYTES_USED + nodes.ramBytesUsed() + scores.ramBytesUsed();
```

Comment:
I really think this is the wrong way to do ramBytesUsed. I am gonna update Lucene as I think there will be measurable performance impact overtime. NeighborArray should simply keep track of underlying array sizes & multiply by the appropriate byte sizes.